### DISCUSSION TITLE
Fragments in 2-level bullet list (non-standard)

### DISCUSSION THREAD

**Discussion by tlyim** [quarto-dev/quarto-cli#1459] [Q&A] [answered]
It looks like that the current implementation of fragments does not work inside a bullet list. 

For example, below will break a bullet list:

```
- item 1.0
::: {.fragment}
  - item 1.1
:::
  - item 1.2
```

And below will always show the leading bullet before the content of that bullet point is revealed next:

```
- item 1.0
  - [item 1.1]{.fragment}
  - item 1.2
```

Is there a simple way to achieve non-standard fragments in a 2-level bullet list like the minimum example below? ({.incremental} cannot do this)

https://user-images.githubusercontent.com/10102234/179799150-65c7417b-0304-467a-8638-ed1af65ec1f2.mp4

**Comment by mcanouil** [comment, marked as answer]
Tricky indentation and empty lines for the first ^^

1. With one fragment

   ````
   ---
   title: "Untitled"
   format: revealjs
   ---
 
   ## Title
 
   - item 1.0
 
     ::: {.fragment}
     - item 1.1
     :::
 
     - item 1.2
 
   ````
 
   https://user-images.githubusercontent.com/8896044/179802661-41239cd9-12a6-4ce0-8a4c-0ccf9834704f.mov

2. With two fragments

   ````
   ---
   title: "Untitled"
   format: revealjs
   ---
 
   ## Title
 
   - item 1.0
 
     ::: {.fragment}
     - item 1.1
     :::
 
     ::: {.fragment}
     - item 1.2
     :::
 
   ````
 
   https://user-images.githubusercontent.com/8896044/179802911-d8262509-8a80-4b4e-b42e-c5391173289b.mov

**Comment by tlyim** [comment]
@mcanouil Thanks for the answer. To use `::: {.fragment}`, there must be an empty line before it. This also introduces an extra line before the bullet list. Is there not a way to avoid this unintended distortion while using `::: {.fragment}`?

```
## Minimum Example

- item 1.0
  - item 1.1
  - item 1.2
- item 1.0
  - item 1.1
  - item 1.2

::: {.fragment}
- item 1.0

  ::: {.fragment}
  - item 1.1
  :::
  
  ::: {.fragment}
  - item 1.2
  :::
::: 
```
![ExtraLine](https://user-images.githubusercontent.com/10102234/180485592-32227614-d40a-4013-9f16-cb5bafc010d6.png)

**Reply by mcanouil** [reply]
I am not sure it's possible since the fragment class is added by encapsulating the element in a `div` tag, thus this breaks the `ul` tag.
<img width="1499" alt="image" src="https://user-images.githubusercontent.com/8896044/180488052-62254721-dc25-4164-9bd1-844f417a6e20.png">

**Reply by mcanouil** [reply]
I found a hacky way (it's likely there is better way) in the right column.

<details><summary>See the code.</summary>

````
---
title: "Habits"
author: "John Doe"
format: revealjs
---

## Minimum Example
::: {.columns}

::: {.column width="50%"}

- item 1.0
  - item 1.1
  - item 1.2

- item 1.0
  - item 1.1
  - item 1.2

::: {.fragment}
- item 1.0

  ::: {.fragment}
  - item 1.1
  :::
  
  ::: {.fragment}
  - item 1.2
  :::
:::

:::

::: {.column width="50%"}

<ul>
- item 1.0
  - item 1.1
  - item 1.2

- item 1.0
  - item 1.1
  - item 1.2

::: {.fragment}
- item 1.0

  ::: {.fragment}
  - item 1.1
  :::
  
  ::: {.fragment}
  - item 1.2
  :::
:::
</ul>

:::

:::

````

</details>

<img width="619" alt="image" src="https://user-images.githubusercontent.com/8896044/180489765-e2b23211-8a38-4dff-977b-1997af02fc8d.png">

**Reply by mcanouil** [reply]
Another solution but not really local is to overwrite the css attributes adding the bottom margin at the end of a list.

````
---
title: "Habits"
author: "John Doe"
format: revealjs
include-after-body:
  text: |
    <style>
    .reveal .slide ul {
      margin-bottom: 0;
    }
    </style>
---

## Minimum Example

- item 1.0
  - item 1.1
  - item 1.2

- item 1.0
  - item 1.1
  - item 1.2

::: {.fragment}
- item 1.0

  ::: {.fragment}
  - item 1.1
  :::
  
  ::: {.fragment}
  - item 1.2
  :::
:::
````

**Reply by tlyim** [reply]
@mcanouil  Thank you for suggesting the workarounds. Very helpful. Much appreciated.

**Comment by jimjam-slam** [comment]
Two other solutions:

If you want to reveal _every_ item in the list in turn, use `.incremental` on the list:

````
::: {.incremental}
- Bananas
- Apples
  - Red apples
  - Green apples
- Grapes
- Kiwi fruit
:::
````

You lose control over precise order of reveals here, but it's quick.

Another is that you can apply `.fragment` to spans, not just divs:

````
::: {.myclass}
- [Bananas]{.fragment}
- [Apples]{.fragment}
  - [Red apples]{.fragment}
  - [Green apples]{.fragment}
- [Grapes]{.fragment}
- [Kiwi fruit]{.fragment}
:::
````

The downside of this approach is that the bullets will still be visible. You can't hide them by applying `list-style-type: none` to the surrounding div because they're already styled, but you could add some CSS targeting `.reveal .myclass ul` or similar (where `.myclass` is on your parent div) to remove the bullets.
